### ISSUE TITLE
Dialogue for the End of the Season

### ISSUE THREAD

**Issue by rdstern** [africanmathsinitiative/R-Instat#2775]
Again this can be modelled on the existing start of the rains!

1. There is an extra field labelled "Evaporation".  Eventually this will have an option to pick up a column - usually from another sheet.  For now it is just a number and 5 is the default.  Could add the extra lebel "mm" if you wish.
2. There is an extra field somewhere with the label "Capacity".  This can be an up-down control - as in the existing Instat.  Also could have an additional label after this field with "mm".
3. Then (only if very easy) ideally it would permit the day chosen to be either the first day < a value for the end, or > value for the start.  Default is less than, for the end.
4. The reason the start is not important - but will be in the future - is that ideally it would then (as in your existing start) allow for a dry spell after the start as well!  This is what we could do in the new improved general system, which is still to come!
5. Have you solved the missing value issue yet?  If not, then my suggestion is as follows - hope it isn't to difficult.  
a) You calculate so when the rainfall is missing, then the water-balance is missing.
b) After any missing you "start again" but with two alternative columns, i.e. first with the water balance on the last missing day set to empty (zero) and the other with it set to full (e.g. 120mm) if that is the capacity.  
c) You have a third column - which is the one you will use.  This is the same as the first column, but is also set to missing if the second column is different to the first one.

That is almost all!  But see the comment below as well!

**Comment by rdstern** [comment]
The one (powerful) extra is to be able to start the calculation for the end on the day of the end of the rains, calculated in the previous dialogue.  This replaces the initial day number by a column taken from the sheet where it was.

I don't think that should be too difficult.  But it may take too long to be done this week.

**Comment by lilyclements** [comment]
I can see how your suggesting to tackling the water balance problem can be done and know how I can implement this in R. However, if on step c) then it is an NA value, surely this then still means the resulting few years will still all be NA so it doesn't help solve that problem?

**Comment by volloholic** [comment]
@lilyclements Is this done now? can I close?

**Comment by lilyclements** [comment]
End of Season needs a small change - it has the option for the evaporation column, but currently this is disabled so the user can only write in the number

End of Rains needs a few changes still, but the R-Code is written in so it's (hopefully!) just little VB-bits

**Comment by rdstern** [comment]
Now that we can use data from a linked worksheet, this end-of-the season dialogue can be "completed".
1. The new "system" for choosing the day numbers can be added - as it has been for the summary, extremes and spells.  This will enable the end of the season the start on the date of the end of the rains.
2. The evaporation should be able to pick-up data on the evaporation from another column in the same (daily) worksheet.  Once it can be used from the daily worksheet that would already be very useful, because any alternative for the evaporation data could be merged easily, before being used.  The function would then have the trivial change of subtracting that value, rather than the fixed value.  
3. If it is easy, then the calculation should cope with missing values in the evaporation column.  But if that is not easy, then I would prefer to make quick progress by checking and mot permitting missing values in the evaporation column.
4. The evaporation will often be available in a linked worksheet.  It MUST be linked so R-Instat knows how to use it, i.e. through the linked columns.  It is an example where the two data frames may be independent and the lin king would then be done manually.
5. Two examples are 1) where the evaporation is available in a data frame of length 366 with the day-of-year being the link. 2) Just 12 values (or 36), where the evaporation is provided for each month, or each 10-day period.  These could then be merged, but ideally this would be an example where we make use of our linking and do the calculations anyway, with (presumably) doing a merge "behind-the-scenes".

**Comment by rdstern** [comment]
There is a problem with the current end of season dialogue.  It doesn't take account of missing values.  This is important.  Currently it seems to put the end on the earliest day if there are missing values.
It isn't easy to take account of them properly, but I had lengthy discussions with Lily, so I guess code is there, but perhaps commented out for now.
It is a problem on this workshop.

**Comment by shadrackkibet** [comment]
@dannyparsons  to fix **2** above. Does this involve just passing **evaporation column** where we have 5 here `(pmax(x + y- 5, 0)` or there is something else that needs to be added to this calculation? 

`water_balance_min <- instat_calculation$new(type="calculation", function_exp="Reduce(function(x, y) pmin(pmax(x + y- 5, 0), 60), replace_NA_min, accumulate=TRUE)", result_name="water_balance_min", sub_calculations=list(sub1=replace_NA_min))`

 and

`water_balance_max <- instat_calculation$new(type="calculation", function_exp="Reduce(function(x, y) pmin(pmax(x + y - 5, 0), 60), replace_NA_max, accumulate=TRUE)", result_name="water_balance_max", sub_calculations=list(sub1=replace_NA_max))`

**Comment by rdstern** [comment]
This depends what "problem" you are fixing!
1. Currently the water balance just uses a constant as you say, e.g. y-5.  We would like to be able to use a column.  This could be a column in the same worksheet, i.e. on a daily basis (easiest initially I assume) or in a linked worksheet.  If in a linked worksheet, then I assume that the evaporation column will be merged, and so will be also daily - as is the rainfall, i.e. x in your formulae above.
2. Currently the end of the season ignores missing values in the data.  It seems to give the end of the season on the same day as was defined as the earliest day, if there are missing values.
3. The idea above was that when the rainfall is missing, then the water balance is set to missing.  The dat after missing rainfall data the extremes are to start with an empty water balance (water_balance_min) or full (water_balance_max).  Then both calculations proceed in parallel until they give the same value.  Until then the water balance is set to missing.  Then - in the calculations - if ever the water balance is missing during the period for the calculations, e.g. day 245 (if starting then) , then the result that year is set to missing.

**Comment by rdstern** [comment]
Here are some small improvements that would also help the end of rains/season dialogue:

1. Change the order of the controls.  Swap the End of Rains and End of Season, so the End of Rains is first!
2. In the End of Rains control change **Amount** to **Last Amount**.
3. Add mm after the 3 up-down controls for Last Amount, Capacity and Water Balance.

And just to check that I think there is current on-going work on 2 bigger issues, namely 
4. Improving the Evaporation to be able to take a column.
5. Solving the missing value bug that currently results in the earliest date being given in some yeurs (not all) when the result should be set to missing.

**Comment by rdstern** [comment]
In the problem with missing values for the end of the season this has just resulted in problems in an analysis in Togo.  This is not so simple, but I thought Lily made some progress before she left.  Once her exams are over, perhaps there could be some discussions between her and Beth on how far she got?

**Comment by lilyclements** [comment]
Hi Roger, what is the issue with the missing values? Is it that it puts the end on the earliest day if there are missing values?
I'll have a look through the r code once exams are out of the way to see if there's anything started on this (it was a while ago so I'm struggling to remember! Sorry!).

**Comment by rdstern** [comment]
That's exactly the problem.  I remember we had the seemingly clever idea that after missing values the water balance would resume twice - once starting empty and the second time starting full.  Then the result that is recorded remains missing until those two options give the same figure.  Then it resumes ordinarily.

Then (once that is done) the water balance is recorded as missing if there are any missing values from the earliest date to the zero value.

**Comment by dannyparsons** [comment]
Closing this but can be reopened if issue is not fixed.

**Comment by bethanclarke** [comment]
I’m still working on evaporation but I have this strange bug. Could one of you @maxwellfundi @shadrackkibet take a look at what I’ve done so far please?

**Comment by rdstern** [comment]
This dialogue seems to be progressing well.  It can now (I think) use a separate evaporation dialogue and Shadrack is currently working on the facility to include a date column and the logical column.
One important feature that remains is to be able to use a linked data frame for evaporation, rather than only have the evaporation in another column in the same sheet as the rainfall.

This is a luxury, because we can always merge.  But is is an example of perhaps the proper way to do this.  There are various examples that could be tried then.  For example in CIMH they have a column for a single year, e.g. one with the day of year (1 to 366) and a value for each day.  Also quite common would be a value for each dekad, i.e. just a column of length 36.

**Comment by rdstern** [comment]
I have moved this back to the 0.4.20 milestone as I hink this is nearly ready now?

**Comment by rdstern** [comment]
This dialogue still doesn't have the saving options for the date and the logical columns.  It should parallel the saving for the start of the rains.  I think these have been done somewhere, but I am not sure?

**Comment by shadrackkibet** [comment]
My understanding is that there is only One component that remains on this. That is adding of options to display Date and  Occurrence for end of rains and end of season. This have been reported in another issue #4112 . So one of this should be closed.
